crates/storage/backend/in_memory.rs

```diff
@@ -0,0 +1,216 @@
+            .read()
+            .map_err(|_| StoreError::Custom("Failed to acquire read lock".to_string()))?;
+        Ok(db
+            .get(&self.table_name)
+            .and_then(|table_ref| table_ref.get(key))
+            .cloned())
+    }
+}
+
+pub struct InMemoryRoTx<'a> {
+    backend: &'a RwLock<Database>,
+}
+
+impl<'a> StorageRoTx for InMemoryRoTx<'a> {
+    fn get(&self, table: &str, key: &[u8]) -> Result<Option<Vec<u8>>, StoreError> {
+        let db = self
+            .backend
+            .read()
+            .map_err(|_| StoreError::Custom("Failed to acquire read lock".to_string()))?;
+
+        Ok(db
+            .get(table)
+            .and_then(|table_ref| table_ref.get(key))
+            .cloned())
+    }
+
+    fn prefix_iterator(
+        &self,
+        table: &str,
+        prefix: &[u8],
+    ) -> Result<Box<dyn Iterator<Item = PrefixResult> + '_>, StoreError> {
+        let db = self
+            .backend
+            .read()
+            .map_err(|_| StoreError::Custom("Failed to acquire read lock".to_string()))?;
+
+        let table_data = db.get(table).cloned().unwrap_or_default();
+        let prefix_vec = prefix.to_vec();
+
+        let results: Vec<PrefixResult> = table_data
+            .into_iter()
+            .filter(|(key, _)| key.starts_with(&prefix_vec))
+            .map(|(k, v)| Ok((k.into_boxed_slice(), v.into_boxed_slice())))
+            .collect();
+
+        let iter = InMemoryPrefixIter {
+            results: results.into_iter(),
+        };
+        Ok(Box::new(iter))
+    }
+}
+
+pub struct InMemoryRwTx {
+    backend: Arc<RwLock<Database>>,
+}
+
+impl StorageRoTx for InMemoryRwTx {
+    fn get(&self, table: &str, key: &[u8]) -> Result<Option<Vec<u8>>, StoreError> {
+        let db = self
+            .backend
+            .read()
+            .map_err(|_| StoreError::Custom("Failed to acquire read lock".to_string()))?;
+
+        Ok(db
+            .get(table)
+            .and_then(|table_ref| table_ref.get(key))
+            .cloned())
+    }
+
+    fn prefix_iterator(
+        &self,
+        table: &str,
+        prefix: &[u8],
+    ) -> Result<Box<dyn Iterator<Item = PrefixResult> + '_>, StoreError> {
+        let db = self
+            .backend
+            .read()
+            .map_err(|_| StoreError::Custom("Failed to acquire read lock".to_string()))?;
+
+        let table_data = db.get(table).cloned().unwrap_or_default();
+        let prefix_vec = prefix.to_vec();
+
+        let results: Vec<PrefixResult> = table_data
+            .into_iter()
+            .filter(|(key, _)| key.starts_with(&prefix_vec))
+            .map(|(k, v)| Ok((k.into_boxed_slice(), v.into_boxed_slice())))
+            .collect();
+
+        let iter = InMemoryPrefixIter {
+            results: results.into_iter(),
+        };
+        Ok(Box::new(iter))
+    }
+}
+
+impl StorageRwTx for InMemoryRwTx {
+    fn put_batch(&mut self, batch: Vec<(&str, Vec<u8>, Vec<u8>)>) -> Result<(), StoreError> {
```

Comment:
This is unnecessarily generic. We should change this to be:


```suggestion
    fn put_batch(&mut self, table: &'static str, batch: Vec<(Vec<u8>, Vec<u8>)>) -> Result<(), StoreError> {
```